File: lib/ts/src/internal/schema/GetSchemaFileMap.ts
//|
//|  Copyright The Telepact Authors
//|  SPDX-License-Identifier: Apache-2.0
//|

import { FsModule, PathModule } from '../../fileSystem.js';
import { SchemaParseFailure } from '../../internal/schema/SchemaParseFailure.js';
import { TelepactSchemaParseError } from '../../TelepactSchemaParseError.js';
import { DocumentLocator, setDocumentLocators } from './DocumentLocators.js';
import { parseTelepactYaml } from './ParseTelepactYaml.js';

export function getSchemaFileMap(directory: string, fs: FsModule, path: PathModule): Record<string, string> {
    const finalJsonDocuments: Record<string, string> = {};
    const documentLocators: Record<string, DocumentLocator> = {};

    const schemaParseFailures: SchemaParseFailure[] = [];

    const paths = fs.readdirSync(directory)
        .sort()
        .map((file) => path.join(directory, file));
    for (const filePath of paths) {
        const relativePath = path.relative(directory, filePath);
        if (fs.statSync(filePath).isDirectory()) {
            schemaParseFailures.push(new SchemaParseFailure(relativePath, [], 'DirectoryDisallowed', {}));
            finalJsonDocuments[relativePath] = '[]';
            continue;
        }

        const content = fs.readFileSync(filePath, 'utf-8');
        if (filePath.endsWith('.telepact.json')) {
            finalJsonDocuments[relativePath] = content;
        } else if (filePath.endsWith('.telepact.yaml')) {
            try {
                const parsed = parseTelepactYaml(content);
                finalJsonDocuments[relativePath] = parsed.canonicalJsonText;
                if (parsed.locator !== undefined) {
                    documentLocators[relativePath] = parsed.locator;
                }
            } catch {
                finalJsonDocuments[relativePath] = '[]';
                schemaParseFailures.push(new SchemaParseFailure(relativePath, [], 'JsonInvalid', {}));
            }
        } else {
            finalJsonDocuments[relativePath] = content;
            schemaParseFailures.push(new SchemaParseFailure(relativePath, [], 'FileNamePatternInvalid', { expected: '*.telepact.json|*.telepact.yaml' }));
        }
    }

    setDocumentLocators(finalJsonDocuments, documentLocators);

    if (schemaParseFailures.length > 0) {
        throw new TelepactSchemaParseError(schemaParseFailures, finalJsonDocuments);
    }

    return finalJsonDocuments;
}

File: lib/go/internal/schema/GetSchemaFileMap.go
//|
//|  Copyright The Telepact Authors
//|  SPDX-License-Identifier: Apache-2.0
//|

package schema

import (
	"os"
	"path/filepath"
	"sort"
	"strings"
)

// GetSchemaFileMap reads Telepact schema resources beneath directory and returns their contents, along with any schema parse failures.
func GetSchemaFileMap(directory string) (map[string]string, []*SchemaParseFailure, error) {
	finalJSONDocuments := make(map[string]string)
	documentLocators := make(map[string]DocumentLocator)
	parseFailures := make([]*SchemaParseFailure, 0)

	entries, err := os.ReadDir(directory)
	if err != nil {
		return nil, nil, err
	}

	sort.Slice(entries, func(i, j int) bool {
		return entries[i].Name() < entries[j].Name()
	})

	for _, entry := range entries {
		relativePath := entry.Name()
		path := filepath.Join(directory, relativePath)

		if entry.IsDir() {
			parseFailures = append(parseFailures, NewSchemaParseFailure(relativePath, nil, "DirectoryDisallowed", map[string]any{}))
			finalJSONDocuments[relativePath] = "[]"
			continue
		}

		content, err := os.ReadFile(path)
		if err != nil {
			return nil, nil, err
		}

		switch {
		case strings.HasSuffix(relativePath, ".telepact.json"):
			finalJSONDocuments[relativePath] = string(content)
		case strings.HasSuffix(relativePath, ".telepact.yaml"):
			canonicalJSON, locator, err := ParseTelepactYAML(string(content))
			if err != nil {
				finalJSONDocuments[relativePath] = "[]"
				parseFailures = append(parseFailures, NewSchemaParseFailure(relativePath, nil, "JsonInvalid", map[string]any{}))
			} else {
				finalJSONDocuments[relativePath] = canonicalJSON
				documentLocators[relativePath] = locator
			}
		default:
			parseFailures = append(parseFailures, NewSchemaParseFailure(relativePath, nil, "FileNamePatternInvalid", map[string]any{"expected": "*.telepact.json|*.telepact.yaml"}))
			finalJSONDocuments[relativePath] = string(content)
		}
	}

	SetDocumentLocators(finalJSONDocuments, documentLocators)
	return finalJSONDocuments, parseFailures, nil
}
